pkg/messageprocessors/devicerepository/devicerepository.go

```diff
@@ -94,7 +95,14 @@ func cacheKey(codec codecType, version *ttnpb.EndDeviceVersionIdentifiers) strin
 	return fmt.Sprintf("%s:%s:%s:%s:%v", version.BrandID, version.ModelID, version.FirmwareVersion, version.BandID, codec)
 }
 
+var (
+	errNoVersionIdentifiers = errors.DefineInvalidArgument("no_version_identifiers", "no version identifiers for device")
+)
+
 func (h *host) retrieve(ctx context.Context, codec codecType, ids ttnpb.ApplicationIdentifiers, version *ttnpb.EndDeviceVersionIdentifiers) (*ttnpb.MessagePayloadFormatter, error) {
+	if version == nil {
+		return nil, errNoVersionIdentifiers.New()
```

Comment:
It should be an error when the device payload formatter is `PayloadFormatter_REPOSITORY` but the device version identifiers are `nil`.